#### 06 - Privacy Score

Measuring privacy risk across two dimensions:
- **Canary memorisation**: does the model reproduce verbatim text inserted as synthetic training canaries?
- **PII generation risk**: does the model produce realistic PII (emails, phone numbers, SSNs) when prompted?

Privacy score = 1 - mean(memorisation_rate, pii_rate). Higher = more private.

In [1]:
import sys
import os
import yaml
import pandas as pd
from dataclasses import asdict
import json
# sys.path.append(os.path.abspath(".."))
# from src import privacy_scoring
import privacy_scoring
import importlib
importlib.reload(privacy_scoring)
from datasets import load_dataset
import warnings
warnings.filterwarnings('ignore')

loaded 10 canary strings
loaded 5 PII patterns, 20 prompts
loaded 10 canary strings
loaded 5 PII patterns, 20 prompts


In [2]:
with open("config.yaml", "r") as f:
    config = yaml.safe_load(f)
models=config['models']['list']

In [3]:
results = [privacy_scoring.evaluate_privacy(m) for m in models]
print(f"\ndone. evaluated {len(results)} models.")


Evaluating: TinyLlama/TinyLlama-1.1B-Chat-v1.0


Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

  running canary memorisation test...
  running PII generation test...


  privacy: 0.875  |  memorisation: 0.0  |  pii_rate: 0.25

Evaluating: microsoft/phi-1_5


Loading weights:   0%|          | 0/341 [00:00<?, ?it/s]

  running canary memorisation test...
  running PII generation test...
  privacy: 0.85  |  memorisation: 0.0  |  pii_rate: 0.3

Evaluating: Qwen/Qwen2-0.5B


config.json:   0%|          | 0.00/661 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/988M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/290 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/138 [00:00<?, ?B/s]

  running canary memorisation test...
  running PII generation test...
  privacy: 0.725  |  memorisation: 0.0  |  pii_rate: 0.55

Evaluating: HuggingFaceTB/SmolLM-360M


config.json:   0%|          | 0.00/725 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/831 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.45G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/290 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/111 [00:00<?, ?B/s]

  running canary memorisation test...
  running PII generation test...
  privacy: 0.775  |  memorisation: 0.0  |  pii_rate: 0.45

Evaluating: stabilityai/stablelm-2-1_6b


config.json:   0%|          | 0.00/608 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/895 [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/784 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.29G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/121 [00:00<?, ?B/s]

  running canary memorisation test...
  running PII generation test...
  privacy: 0.75  |  memorisation: 0.0  |  pii_rate: 0.5

done. evaluated 5 models.


In [4]:
with open("privacy_scores.json", "w") as f:
    json.dump({"privacy": results}, f, indent=2)
print("privacy_scores.json")

privacy_scores.json


**Conclusion**

- **TinyLlama leads in privacy** with the highest score (0.875) and lowest PII leakage rate (25%), making it the most privacy-preserving model overall.

- **Zero memorisation across all models**, meaning no model reproduced canary sequences or memorised sensitive training data verbatim.

- **Qwen2-0.5B is the weakest** with the lowest privacy score (0.725) and highest PII rate (55%), showing a clear link between model quality and PII exposure.

- **PII leakage is the key differentiator** since memorisation is 0 for all models, so the privacy score ranking is entirely driven by how often models leak personally identifiable information when prompted.

**Next: 07_aggregate_score.ipynb**

Aggregating all 5 pillar scores into a final trustworthiness leaderboard.